packages/flutter_migrate/lib/src/compute.dart

```diff
@@ -0,0 +1,1014 @@
+// found in the LICENSE file.
+
+import 'base/common.dart';
+import 'base/file_system.dart';
+import 'base/logger.dart';
+import 'base/project.dart';
+import 'base/terminal.dart';
+import 'custom_merge.dart';
+import 'environment.dart';
+import 'flutter_project_metadata.dart';
+import 'result.dart';
+import 'utils.dart';
+
+// This defines files and directories that should be skipped regardless
+// of gitignore and config settings. Using `/` as a stand-in for path separator.
+const List<String> _skippedFiles = <String>[
+  'ios/Runner.xcodeproj/project.pbxproj', // Xcode managed configs that may not merge cleanly.
+  'README.md', // changes to this shouldn't be overwritten since is is user owned.
+];
+
+const List<String> _skippedDirectories = <String>[
+  '.dart_tool', // ignore the .dart_tool generated dir
+  '.git', // ignore the git metadata
+  'lib', // Files here are always user owned and we don't want to overwrite their apps.
+  'test', // Files here are typically user owned and flutter-side changes are not relevant.
+  'assets', // Common directory for user assets.
+  'build' // Build artifacts
+];
+
+bool _skipped(String localPath, FileSystem fileSystem,
+    {Set<String>? blacklistPrefixes}) {
+  for (final String path in _skippedFiles) {
+    if (path.replaceAll('/', fileSystem.path.separator) == localPath) {
+      return true;
+    }
+  }
+  if (_skippedFiles.contains(localPath)) {
+    return true;
+  }
+  for (final String dir in _skippedDirectories) {
+    if (localPath.startsWith(
+        '${dir.replaceAll('/', fileSystem.path.separator)}${fileSystem.path.separator}')) {
+      return true;
+    }
+  }
+  if (blacklistPrefixes != null) {
+    for (final String prefix in blacklistPrefixes) {
+      if (localPath.startsWith('$prefix${fileSystem.path.separator}')) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+const List<String> _skippedMergeFileExt = <String>[
+  // Don't merge image files
+  '.png',
+  '.jpg',
+  '.jpeg',
+  '.gif',
+  '.bmp',
+  '.svg',
+  // Don't merge compiled artifacts and executables
+  '.jar',
+  '.so',
+  '.exe',
+];
+
+const Set<String> _alwaysMigrateFiles = <String>{
+  '.metadata', // .metadata tracks key migration information.
+  'android/gradle/wrapper/gradle-wrapper.jar',
+  // Always add .gitignore back in even if user-deleted as it makes it
+  // difficult to migrate in the future and the migrate tool enforces git
+  // usage.
+  '.gitignore',
+};
+
+/// True for files that should not be merged. Typically, images and binary files.
+bool _skippedMerge(String localPath) {
+  for (final String ext in _skippedMergeFileExt) {
+    if (localPath.endsWith(ext) && !_alwaysMigrateFiles.contains(localPath)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/// Data class holds the common context that is used throughout the steps of a migrate computation.
+class MigrateContext {
+  MigrateContext({
+    required this.migrateResult,
```

Comment:
Presumably this one is mutable? Separating mutable state and immutable state into two separate objects, one of which is marked as immutable, would make it a lot easier to reason about what's being handed around and what can happen to it.